core/src/main/java/org/apache/iceberg/ManifestGroup.java

```diff
@@ -204,4 +141,78 @@ public ManifestGroup select(String... columnNames) {
 
     return CloseableIterable.concat(readers);
   }
+
+  static Builder builder(FileIO io, Iterable<ManifestFile> manifests) {
+    return new Builder(io, manifests);
+  }
+
+  static class Builder {
```

Comment:
I'm undecided about the change to use a builder instead of a refinement API. The builder is simpler, but the implementation ends up doing very little and there is for some reason a `build()` call in the chain required just before `entries()`.

On the other hand, the refinement API is a bit hard to maintain and is a little confusing. Its purpose is to make the ManifestGroup immutable so it can be shared, but that seems unlikely. We can wrap it in refinement APIs if we want to do this.

How about a third option: let's make `ManifestGroup` mutable. That way it is essentially a builder that you don't need to call `build()` on. You can just call `entries()` directly. The advantage is that the API remains the same (fewer changes everywhere) and we have a simpler class to maintain because it doesn't need to follow the refinement pattern.

What do you think?

Reply:
Yea, the refinement api was getting confusing for me. I'm ok with either builder vs mutable class. I'll change to a mutable version if that's easier.

Reply:
One problem with the mutable approach is now we have to be extra careful with the instantiation of  `ManifestGroup`. For instance, `evalCache` requires parameters `partitionFilter`,  `dataFilter` and `specById`.  When setting any one of these parameters we'd have to initialized `evalCache` again. 

I think I'd rather keep the builder here.

Reply:
What if we make the cache a local variable instead of sharing it across invocations of `entries`? Since we don't expect multiple invocations, it should have identical performance. We don't really need to keep a cache around.